### ISSUE TITLE
Wrong order of generated JS code

### ISSUE THREAD

**Issue by vadwook** [elm-lang/elm-compiler#1341]
I stumbled over this issue during my development. Provided Elm code is dumbed down version of what I was working on so it might looks _nonsensical_:
```elm
import Html exposing (Html, text)
import List exposing (filter, head)
import Maybe


pageIndex : String -> Html
pageIndex attr =
    text "Main Page"

page1 : String -> Html
page1 attr =
    text (first "NOTHING")

notFound : String -> Html
notFound attr =
    text "Not Found"


pages : List (String -> Html, String)
pages = 
    [ (pageIndex, "pageIndex")
    , (page1, "page1")
    ]

first : String -> String
first alter =
    case head pages of
        Just p -> snd p
        Nothing -> alter

view : Html
view =
    let
        filteredPages name = 
            filter (\p -> snd p == name) pages
        v = fst (Maybe.withDefault (notFound, "notFound") (head (filteredPages "page1")))
    in
        v "cat"

main : Html
main =
    view
```
Assuming we have above code in file named `BadMain.elm`:
`elm make BadMain.elm --output bad_index.html`

Executing `elm make` does not gives any errors **but** when you run `bad_index.html` in your web browser you will get:

> TypeError: v is not a function

I looked into generated JS code and debugged it a little. I found this particular code to be wrong:
```javascript
var notFound = function (attr) {    return $Html.text("Not Found");};
var pageIndex = function (attr) {    return $Html.text("Main Page");};
var pages = _U.list([{ctor: "_Tuple2",_0: pageIndex,_1: "pageIndex"},{ctor: "_Tuple2",_0: page1,_1: "page1"}]);
var page1 = function (attr) {    return $Html.text(first("NOTHING"));};
var first = function (alter) {    var _p0 = $List.head(pages);if (_p0.ctor === "Just") {    return $Basics.snd(_p0._0);} else {    return alter;}};
```
`var page1` is declared after `pages` but `pages` references `page1`. This will result in storing `undefined` in `pages` and breaking things later one.

Additionally, I noticed that when you change definition of `page1` to:
```elm
page1 : String -> Html
page1 attr =
    text "NOTHING"
```
JS code in `bad_index.html` will be generated in correct order and everything will work correctly.

Used Elm version: **elm-make 0.16 (Elm Platform 0.16.0)**

**Comment by jvoigtlaender** [comment]
This is just another instance of #873. Maybe the twentieth or thirtieth that got reported. :smile:

You can safely close the issue here, just as all the others get closed until the "mother issue" #873 will eventually be resolved.

**Comment by vadwook** [comment]
Very well then, closing as _specific_ case of #873.
